consensus/dbft/dkg.go

```diff
@@ -340,275 +340,315 @@ func (c *DBFT) handleDKG(snapshot *snapshot, keystore *antimev.KeyStore, h *type
 
-// loopWatchTaskList retries every transaction task in transaction watch list.
-func (c *DBFT) loopWatchTaskList() {
-	log.Info("DKG transaction dispatcher started")
-	defer log.Info("DKG transaction dispatcher stopped")
+// dkgTaskWatcher retries every transaction task in transaction watch list.
+func (c *DBFT) dkgTaskWatcher() {
+	log.Info("DKG task watcher started")
 
 	var watchTaskList []*task
-	for range c.loopWatchTaskChan {
-		currentHeight := (*c.backend).CurrentBlock().Number.Uint64()
-		if len(c.appendWatchTaskChan) > 0 {
-			// Append watchList from appendWatchTaskChan to watchTaskList.
-			newWatchList := <-c.appendWatchTaskChan
-			watchTaskList = append(watchTaskList, *newWatchList...)
-		}
-		log.Info("DKG loopWatchTaskList", "currentHeight", currentHeight, "watchListLength", len(watchTaskList))
-
-		// Loop tasks in watchTaskList.
-		var retryList []*task
-		if len(watchTaskList) > 0 {
-			for _, item := range watchTaskList {
-				if currentHeight < item.EndHeight && !item.ConfirmedSuccess {
-					needRetry := false
-					// Send failed, just resend and set transaction hash.
-					if item.TxHash == nil {
-						needRetry = true
-					}
+watchLoop:
+	for {
+		select {
+		case <-c.quit:
+			break watchLoop
+		case newWatchList := <-c.dkgTaskWatcherCh:
+			if newWatchList != nil && len(*newWatchList) > 0 {
+				// Append watchList from appendWatchTaskChan to watchTaskList.
+				watchTaskList = append(watchTaskList, *newWatchList...)
+			}
+			currentHeight := (*c.backend).CurrentBlock().Number.Uint64()
+			log.Info("DKG task watcher", "currentHeight", currentHeight, "watchListLength", len(watchTaskList))
 
-					// Send successfully, wait 3 blocks to check transaction status.
-					if item.TxHash != nil && currentHeight-item.SendHeight == 3 {
-						receipt, err := c.txAPI.GetTransactionReceipt(context.Background(), *item.TxHash)
-						if err != nil {
+			// Loop tasks in watchTaskList.
+			var retryList []*task
+			if len(watchTaskList) > 0 {
+				for _, item := range watchTaskList {
+					if currentHeight < item.EndHeight && !item.ConfirmedSuccess {
+						needRetry := false
+						// Send failed, just resend and set transaction hash.
+						if item.TxHash == nil {
 							needRetry = true
-							log.Error("DKG get transaction receipt failed", "err", err, "txHash", item.TxHash)
 						}
-						if receipt["status"] != types.ReceiptStatusSuccessful {
-							needRetry = true
-							log.Error("DKG get transaction receipt status error", "txHash", item.TxHash, "status", receipt["status"])
+
+						// Send successfully, wait 3 blocks to check transaction status.
+						if item.TxHash != nil && currentHeight-item.SendHeight == 3 {
+							receipt, err := c.txAPI.GetTransactionReceipt(context.Background(), *item.TxHash)
+							if err != nil {
+								needRetry = true
+								log.Error("DKG get transaction receipt failed", "err", err, "txHash", item.TxHash)
+							}
+							if receipt["status"] != types.ReceiptStatusSuccessful {
```

Comment:
If the call to `GetTransactionReceipt()` fails and returns an error, then the `receipt` variable will be nil. Attempting to access it directly will cause a panic.

```suggestion
							} else if receipt["status"] != types.ReceiptStatusSuccessful {
```

Reply:
Seems like an issue not related to this PR but the original #332, we may open another PR to fix this.

But the fix is simple, yes, we can commit your suggestion now as well.

Reply:
Yes, I just happened to see this code and noticed a minor issue here. So I commented on it right there.